docs: describe OT2 architecture (#6245)
Opentrons/opentrons: OT2_ARCHITECTURE.md

# OT2 Architecture

## Hardware

### Raspberry Pi
  
Inside the OT2 is a small computer (a [Raspberry Pi 3 Model B](https://www.raspberrypi.org/products/raspberry-pi-3-model-b-plus/)), running Linux.

> Data is stored on the Raspberry Pi's SD card. The majority of the filesystem is read-only but certain directories such as `/data/` and `/var/lib/jupyter/notebooks` allow read-write access. The OT2 camera (a webcam) is connected to a USB port on the Raspberry Pi (this port, and a 4th port which may have an unused thumbdrive, are normally covered by the Pi's case).

Author: Technically this of course make more sense in the Software section - I might leave that for a later refactor from you folks as at the moment I think this flows OK, even though I know you're right.

#### Networking

> Other computers, like your laptop, communicate with the Raspberry Pi that controls the OT2 via a network. When you connect your robot via USB, you are in fact connecting to the Raspberry Pi's ethernet port. The USB socket on the robot is connected to an internal USB-to-ethernet adapter, which is in turn connected to the Pi's ethernet port. When you connect a cable to the robot's USB port, your computer recognises the device as a new network interface, which establishes a local network with the OT2. (You can also connect the OT2 to a wireless network with WiFi, or connect to a wired network by placing a USB-ethernet adapter in one of the Pi's USB ports).

Reviewer: ```suggestion
Your computer communicates with the Raspberry Pi inside the OT-2 via a network connection.

When you connect to your robot via USB, you're actually connecting to the Raspberry Pi's Ethernet port. The USB socket on the side of the robot leads to an internal USB-to-Ethernet adapter, which in turn leads to the Pi's Ethernet port.

When you connect your computer to the robot's USB port, your computer recognizes the device as a new network interface, and establishes a [link-local](https://en.wikipedia.org/wiki/Link-local_address) connection to the OT-2.

(You can also connect the OT-2 to a Wi-Fi network.)
```

Reviewer: * Split into paragraphs for readability.
* Tiny readability tweaks.
* Deleted the tip about connecting to a wired network by placing a USB-Ethernet adapter in one of the Pi's USB ports. I know we suggested this in https://github.com/Opentrons/opentrons/issues/2836#issuecomment-548443541, but I don't know if I'm comfortable documenting it right now.

  Opentrons Support used to show customers how to crack open the OT-2's electronics enclosure and plug into the Pi's native Ethernet port. But an anecdote from the Hardware Testing team suggested that chatty Ethernet networks could cause brownouts on some robots. Would a USB-powered Ethernet adapter work around that? I don't know. Maybe, maybe not.
  
  Upcoming hardware revisions will replace the OT-2's side USB port with a side Ethernet port, which I'm very excited for.

  Until then, if someone wants an Ethernet connection enough to specifically ask Opentrons Support about it, we'll be happy to give them the whole story. But we won't advertise the capability in documentation.

  How does that sound to you?

Author: Really pleased that an ethernet port is coming! IMO this is a pretty key feature. All of my robots have used ethernet connections and I think it's kind of a must for reliability in many applications. My feeling would be to put this in with whatever caveats are necessary because I don't thin kit's something people would necessarily ask about even if it could be useful, but I'll defer atm!

### Lights, switches

The Raspberry Pi controls the lights of the robot, and detects presses on the switch and opening of the door via its General Purpose Input-Output (GPIO) ports. The GPIO ports also allow the Raspberry Pi to reset the motor controller board (described below).

### Motor control

The robot has 6 axes of motion, each controlled by a stepper motor:
  

 - X: left/right
 - Y: forward/back
 - Z: left pipette up/down
 - A: right pipette up/down
 - B: left pipette plunger
 - C: right pipette plunger

  

> The motors of the robot, and the endstops and probing switches that help position them, are connected to the motor controller board, housed in the gantry, which is a [SmoothieBoard](http://smoothieware.org/smoothieboard) running  [a custom fork of Smoothieware](https://github.com/Opentrons/SmoothiewareOT).

Reviewer: ```suggestion
The motors, endstops, and probing switches are connected to the motor controller board, housed in the gantry. The motor controller board is a modified [Smoothieboard](http://smoothieware.org/smoothieboard) running [a custom fork of Smoothieware](https://github.com/Opentrons/SmoothiewareOT).
```

* Shorter sentences, for readability.
* In addition to the motor controller board using customized *firmware,* it's actually also customized hardware. I forget exactly what the changes are, but I don't think a stock Smoothieboard is drop-in compatible, unfortunately.

> The SmoothieBoard is connected to the Raspberry Pi via a UART connection, with the Pi sending commands in [GCODE](https://en.wikipedia.org/wiki/G-code).

Reviewer: ```suggestion
The motor controller board is connected to the Raspberry Pi via a UART connection, with the Pi sending commands in [G-code](https://en.wikipedia.org/wiki/G-code).
```

In part due to the customized hardware mentioned in my suggestion above, we try to consistently call it the "motor controller board" instead of the "SmoothieBoard" nowadays.

# OT2 Architecture

## Hardware

### Raspberry Pi
  
Inside the OT2 is a small computer (a [Raspberry Pi 3 Model B](https://www.raspberrypi.org/products/raspberry-pi-3-model-b-plus/)), running Linux.

#### Data storage

> Data on the Raspberry Pi is stored in two physical pieces of hardware. The main operating system of the robot is stored on an SD card in the Raspberry Pi. This SD card is read-only apart from when the robot is being updated. Customized user data is stored on a thumb-drive in one of the USB ports of the Raspberry Pi (normally the case covers this USB port so it is not visible). The OT2 camera (a webcam) is connected to another normally hidden USB port of the Raspberry Pi.

Reviewer 2: We actually don't use the USB thumb drive anymore - they tended to be unreliable, and the presence of the USB thumb drive was to work around Balena invalidating the SD card data in certain circumstances, which we solved by moving away from Balena.

Author: Awesome - fixed. If you have time at some point to explain where what is read only and what is not is defined that is something that I'm still not clear on.

#### Networking

  Other computers, like your laptop, communicate with the Raspberry Pi that controls the OT2 via a network. When you connect your robot via USB, you are in fact connecting to the Raspberry Pi's ethernet port. The USB port is connected to an internal USB-to-ethernet adapter, which is in turn connected to the Pi's ethernet port. When you connect a cable to the robot's USB port, your computer recognises the device as a new network interface, which establishes a local network with the OT2. (You can also connect the OT2 to a wireless network with WiFi, or connect to a wired network by placing a USB-ethernet adapter in one of the Pi's USB ports).

### Lights, switches

The Raspberry Pi controls the lights of the robot, and detects presses on the switch and opening of the door via its General Purpose Input-Output (GPIO) ports. The GPIO ports also allow the Raspberry Pi to reset the motor controller board (described below).

### Motor control

The robot has 6 axes of motion, each controlled by a stepper motor:
  

 - X: left-right
 - Y: forward-back
 - Z: left pipette up-down

>  - A: right pipette up-down

Reviewer 2: These mappings are slightly wrong. Z is left pipette up/down, A is right pipette up/down, B is left pipette plunger, C is right pipette plunger

Author: (Thanks for the comments!)
I think that's what this says but I'll change "plunger up-down" to just "plunger"

Note that only one process can have access to the robot's GPIO ports at a time. By default the robot server connects to these ports on start-up, which prevents a copy of the API running in Jupyter or imported from a Python script from using GPIO functionality. To gain access to the GPIO's in these custom scripts, one can disable the robot server with `systemctl stop opentrons-robot-server`, before importing the Opentrons API into the other Python script. However this will prevent the Opentrons app from connecting to the robot until the robot server is restarted with `systemctl start opentrons-robot-server`, or the robot is rebooted.

## Shared data
Some information, such as labware geometry data, needs to be used in multiple locations in the Opentrons codebase (e.g. in the protocol designer as well as on the robot itself). This data is kept in a special [shared data](https://github.com/Opentrons/opentrons/tree/edge/shared-data) repository so that it can be easily included in all these locations.
  

## Update server

> The [update server](https://github.com/Opentrons/opentrons/tree/edge/update-server) is a separate server designed simply to allow updating the robot's software with a new system image built using buildroot.

Reviewer 2: Perhaps worth noting that the update server also controls some system-specific tasks such as handling SSH keys and setting the robot's name.

Author: Done!
